lib/Frontend/ModuleInterfaceSupport.cpp

```diff
@@ -650,7 +650,14 @@ class InheritedProtocolCollector {
         printer << " {}";
       };
 
-      printBody();
+      if (printOptions.PrintCompatibilityFeatureChecks) {
+        Decl *decl = const_cast<Decl *>(cast<Decl>(nominal));
+        printWithCompatibilityFeatureChecks(printer, curPrintOptions, decl,
+                                            [&] { printBody(); });
+      } else {
+        printBody();
+      }
```

Comment:
Okay, so @slavapestov removed this logic in a recent patch because (1) he realized that it's checking based on the features used by the extended declaration and not the extension itself, so it's a bit broken, and (2) it wasn't tested in any way.  If we're going to put this back, we should really fix the first part.  It looks like extensions are considered to use the `$GlobalActors` feature if they're extending a type that's a global actor, so maybe you just need to pass in the extension here.

There's logic in the printing of extensions to suppress the extension if the declaration is

It looks like the general bug here is that we need to consider a feature to be used by an extension if it's necessary in order to declare the extended type.  If we just suppress the extended type in its entirety, we need to suppress extensions of it as well, even if they're

Reply:
There's no ExtensionDecl here though; these extensions are "synthesized", meaning they only exist in the imagination of the ASTPrinter.

Reply:
Can we synthesize a real `ExtensionDecl`, or is that difficult in practice?

Reply:
> Do I understand correctly that what I have currently is correct but theoretically insufficient for fixing some other bugs where the extension, rather than the extended type, depends on a conditionally available feature?

At the very least, I think it probably needs to consider the availability of both the protocol and the extended type.

It will also redundantly emit the extension under different conditions if the nominal type is emitted redundantly that way, which is unnecessary but I guess not actually a problem.

Reply:
My opinion on this is only based on a first impression, not being very experienced with this code yet, but synthesizing  `ExtensionDecl` feels like it could be a maintenance burden down the road. I don't see any other examples of instantiating an extension decl without it actually being in a source file. How about I refactor `printWithCompatibilityFeatureChecks()` a bit such that I can reuse just the part that does the printing with a given `FeatureSet` and then I manually compute the `FeatureSet` as the union of features required by the extended type and the protocol?

Reply:
That's not quite right either; if the protocol uses primary associated types, we don't need to #if the conformance; only the protocol definition itself is wrapped in #if.

Reply:
I see, I read over https://github.com/apple/swift/pull/42238 more carefully and I understand the issue now. So let's suppose I had an extension decl. If I passed it to `printWithCompatibilityFeatureChecks()` instead of the extended type, then eventually the logic in `usesFeatureGlobalActors()` would be invoked and it would, as currently implemented, do exactly the same thing I have here in this PR. I'm also not sure how `usesFeatureGlobalActors()` would be updated to query whether conformance to a protocol implies that a feature guard must be used for the conformance extension.

Passing an extension decl would certainly be more future proof, but unless we have a concrete example of an existing protocol where what I have now is going to break I'd like to leave this as-is and document the issue as a known shortcoming of the approach.

Reply:
I'm taking a deeper look at synthesizing an `ExtensionDecl` first. I did find an example that I can maybe follow.

Reply:
Alright, I took a stab at synthesizing an `ExtensionDecl`. Does this look reasonable to you @rjmccall and @slavapestov? It feels a bit a side-effectful to me.

Reply:
Slava can speak to the request-evaluator parts better than I can.

Reply:
The request evaluator parts look fine, but I'm wondering if you can remove the printBody lambda with it's bespoke printing logic and just feed the synthesized ExtensionDecl to the ASTPrinter itself?

Reply:
I guess you'll need to attach the correct availability and SPI attributes to the synthesized extension also, but it would still be better to not duplicate the printing logic here.

Reply:
It does seem feasible to let the `ASTPrinter` be responsible for the printing. I am having to handle a number of small details with care, though, to ensure that the `swiftinterface` contents remain unchanged after taking this approach. (I don't want this under the hood change to cause a lot of churn in interfaces for no good reason since folks diff the interfaces to understand what is being added and removed from build to build).

Reply:
Alright, I completely revamped this to use the `ASTPrinter` to emit the extension declaration. PTAL @slavapestov